### PR TITLE
KT-5045 Intention to convert between two comparisons and range check and vice versa

### PR DESCRIPTION
**Pull request by shiraji**
#KT-5045 Fixed

https://youtrack.jetbrains.com/issue/KT-5045

### CODE REVIEW

File: idea/src/org/jetbrains/kotlin/idea/intentions/ConvertRangeCheckToTwoComparisonsIntention.kt [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+        val arg = element.left?.text ?: return
+        val max: Any = rangeExpression.right?.text ?: return
+        val comparisonsExpression = KtPsiFactory(element).createExpressionByPattern("$0 <= $1 && $1 <= $2", min, arg, max)
+        element.replace(comparisonsExpression)
+    }
+
+    override fun isApplicableTo(element: KtBinaryExpression): Boolean {
+        if (element.operationToken != KtTokens.IN_KEYWORD) return false
+        // ignore for-loop. for(x in 1..2) should not be convert to for(1<=x && x<=2)
+        if (element.getStrictParentOfType<KtForExpression>() != null) return false
```
Comment (mglukhikh) [inline idea/src/org/jetbrains/kotlin/idea/intentions/ConvertRangeCheckToTwoComparisonsIntention.kt]: I'd say this condition is too strict. E.g. your intention will not work for code like

```
for (element in list) {
    if (element in 1..4) { ... }
}
```

File: idea/src/org/jetbrains/kotlin/idea/intentions/ConvertRangeCheckToTwoComparisonsIntention.kt [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+import org.jetbrains.kotlin.psi.createExpressionByPattern
+import org.jetbrains.kotlin.psi.psiUtil.getStrictParentOfType
+
+class ConvertRangeCheckToTwoComparisonsIntention : SelfTargetingOffsetIndependentIntention<KtBinaryExpression>(KtBinaryExpression::class.java, "Convert to comparisons") {
+    override fun applyTo(element: KtBinaryExpression, editor: Editor?) {
+        if (element.operationToken != KtTokens.IN_KEYWORD) return
+        val rangeExpression = element.right as? KtBinaryExpression ?: return
+        val min = rangeExpression.left?.text ?: return
+        val arg = element.left?.text ?: return
+        val max: Any = rangeExpression.right?.text ?: return
```
Comment (mglukhikh) [inline idea/src/org/jetbrains/kotlin/idea/intentions/ConvertRangeCheckToTwoComparisonsIntention.kt]: Why do you need `Any` here?

File: idea/src/org/jetbrains/kotlin/idea/intentions/ConvertTwoComparisonsToRangeCheckIntention.kt [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+        val firstCondition = condition.left as? KtBinaryExpression ?: return null
+        val secondCondition = condition.right as? KtBinaryExpression ?: return null
+        val firstOpToken = firstCondition.operationToken
+        val secondOpToken = secondCondition.operationToken
+        val firstLeft = firstCondition.left ?: return null
+        val firstRight = firstCondition.right ?: return null
+        val secondLeft = secondCondition.left ?: return null
+        val secondRight = secondCondition.right ?: return null
+
+        return when (firstOpToken) {
```
Comment (mglukhikh) [inline idea/src/org/jetbrains/kotlin/idea/intentions/ConvertTwoComparisonsToRangeCheckIntention.kt]: This `when` is written in a very straight-forward way. Please, think a bit how to make it shorter. All branches are very like one another.

File: idea/src/org/jetbrains/kotlin/idea/intentions/ConvertTwoComparisonsToRangeCheckIntention.kt [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+                        val longText = if (text.endsWith("l") || text.endsWith("L")) text.substring(0, text.length - 1) else text
+                        longText.toLong() - 1
+                    }
+                    KotlinBuiltIns.isShort(type) -> java.lang.Short.parseShort(text) - 1
+                    KotlinBuiltIns.isChar(type) -> text[0].toInt() - 1
+                    else -> return null
+                }
+                return number.toString()
+            }
+            else -> return "($text - 1)"
```
Comment (mglukhikh) [inline idea/src/org/jetbrains/kotlin/idea/intentions/ConvertTwoComparisonsToRangeCheckIntention.kt]: This is very controversial. What if expression is a `String`, for example?

File: idea/src/org/jetbrains/kotlin/idea/intentions/ConvertTwoComparisonsToRangeCheckIntention.kt [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+            is KtConstantExpression -> {
+                val number: Number = when {
+                    KotlinBuiltIns.isInt(type) -> text.toInt() - 1
+                    KotlinBuiltIns.isLong(type) -> {
+                        val text = text
+                        val longText = if (text.endsWith("l") || text.endsWith("L")) text.substring(0, text.length - 1) else text
+                        longText.toLong() - 1
+                    }
+                    KotlinBuiltIns.isShort(type) -> java.lang.Short.parseShort(text) - 1
+                    KotlinBuiltIns.isChar(type) -> text[0].toInt() - 1
```
Comment (mglukhikh) [inline idea/src/org/jetbrains/kotlin/idea/intentions/ConvertTwoComparisonsToRangeCheckIntention.kt]: I think it is better to replace character with character, either explicit e.g. `'z' - 1` or implicit `'y'`

File: idea/src/org/jetbrains/kotlin/idea/intentions/ConvertTwoComparisonsToRangeCheckIntention.kt (hunk 4 above)
Comment (shiraji) [inline idea/src/org/jetbrains/kotlin/idea/intentions/ConvertTwoComparisonsToRangeCheckIntention.kt]: Moved type check inside the method.

File: idea/src/org/jetbrains/kotlin/idea/intentions/ConvertTwoComparisonsToRangeCheckIntention.kt (hunk 3 above)
Comment (shiraji) [inline idea/src/org/jetbrains/kotlin/idea/intentions/ConvertTwoComparisonsToRangeCheckIntention.kt]: I could not come up with better solution...do you have any suggestion?

File: idea/src/org/jetbrains/kotlin/idea/intentions/ConvertRangeCheckToTwoComparisonsIntention.kt (hunk 2 above)
Comment (shiraji) [inline idea/src/org/jetbrains/kotlin/idea/intentions/ConvertRangeCheckToTwoComparisonsIntention.kt]: done by e485cba

File: idea/src/org/jetbrains/kotlin/idea/intentions/ConvertRangeCheckToTwoComparisonsIntention.kt (hunk 1 above)
Comment (shiraji) [inline idea/src/org/jetbrains/kotlin/idea/intentions/ConvertRangeCheckToTwoComparisonsIntention.kt]: done by 4d1adcb

**Comment by mglukhikh** [comment]
Ok, thank you, I will merge it manually. Later (probably tomorrow) you can look at this `when` code, I have simplified it successfully.

**Comment by shiraji** [comment]
Nice. I really wanna know how you can do that. I'll take a look the code. Thanks!

**Comment by mglukhikh** [comment]
Finally merged it. Sorry, too many things to do these days

**Comment by shiraji** [comment]
Thanks. I checked your code. It's very nice. No worry for late. I know these days are busiest days in the year.

**Pull request closed without merging**